app.py

```diff
@@ -0,0 +1,60 @@
+import base64
+import binascii
+import hashlib
+import hmac
+import urllib.parse
+
+from flask import abort, Flask, redirect, request
+
+app = Flask(__name__)
+
+try:
+    SECRET = open('keys/secret').read().strip()
+except FileNotFoundError:
+    print('No secret file found')
+
+def secret_hmac(msg):
+    return hmac.new(SECRET, msg, digestmod=hashlib.sha256).digest()
+
+@app.route("/test")
+def test():
+    try:
+        username = request.headers['X-WEBAUTH-USER'].strip()
+        return "Logged in as {}".format(username)
+    except KeyError:
+        return "No username, something is very wrong!"
+
+@app.route("/")
+def do_it():
+    sso = request.args.get('sso')
+    sig = request.args.get('sig')
+
+    got_sig = secret_hmac(sso.encode())
+
+    if not hmac.compare_digest(binascii.unhexlify(sig), got_sig):
+        abort(404)
+    
+    username = request.headers['X-WEBAUTH-USER'].strip()
+
+    sso_parse = urllib.parse.parse_qs(base64.b64decode(sso).decode())
```

Comment:
Maybe specify `decode('ascii')` just to be clear?

Reply:
Is this guaranteed to be ascii?

Reply:
I would presume so. If it's not ascii, do you know what encoding it will be?

Reply:
The [reference implementation](https://github.com/discourse/discourse/blob/master/lib/single_sign_on.rb) doesn't seem to assume any encoding, so I assume it's Unicode? Though I'm admittedly not well-versed in Ruby idioms...